core/src/test/scala/akka/persistence/cassandra/journal/TagWritersSpec.scala

```diff
@@ -28,29 +30,57 @@ class TagWritersSpec extends TestKit(ActorSystem("TagWriterSpec"))
 
       tagWriters ! TagFlush("blue")
       probe.expectMsg(Flush)
-      probe.reply(Flushed)
-      expectMsg(Flushed) // should be forwarded
+      probe.reply(FlushComplete)
+      expectMsg(FlushComplete) // should be forwarded
     }
 
     "flush all tag writers" in {
-      val probe = TestProbe()
-      val tagWriters = system.actorOf(TagWriters.props((_, tag) => probe.ref))
+      val redProbe = TestProbe()
+      val blueProbe = TestProbe()
+      val probes = Map("red" -> redProbe, "blue" -> blueProbe)
+      val tagWriters = system.actorOf(TagWriters.props((_, tag) => probes(tag).ref))
 
       // do something to make it create a couple
-      val tp = TagProgress("p", 1, 1)
-      tagWriters ! SetTagProgress("blue", tp)
-      probe.expectMsg(SetTagProgress("blue", tp))
-      tagWriters ! SetTagProgress("red", tp)
-      probe.expectMsg(SetTagProgress("red", tp))
+      val blueTagWrite = TagWrite("blue", Vector.empty)
+      tagWriters ! blueTagWrite
+      blueProbe.expectMsg(blueTagWrite)
+      val redTagWrite = TagWrite("red", Vector.empty)
+      tagWriters ! redTagWrite
+      redProbe.expectMsg(redTagWrite)
 
       tagWriters ! FlushAllTagWriters
-      probe.expectMsg(Flush)
-      probe.reply(Flushed)
+      blueProbe.expectMsg(Flush)
+      blueProbe.reply(FlushComplete)
+      redProbe.expectMsg(Flush)
+      // All flushed comes after all have flushes
       expectNoMessage(100.millis)
-      probe.expectMsg(Flush)
-      probe.reply(Flushed)
+      redProbe.reply(FlushComplete)
       expectMsg(AllFlushed)
     }
+
+    "sends blank progress to tag writers not in recovery" in {
+      val redProbe = TestProbe()
+      val blueProbe = TestProbe()
+      val probes = Map("red" -> redProbe, "blue" -> blueProbe)
+      val tagWriters = system.actorOf(TagWriters.props((_, tag) => probes(tag).ref))
+
+      // do something to make it create a couple
+      val blueTagWrite = TagWrite("blue", Vector.empty)
+      tagWriters ! blueTagWrite
+      blueProbe.expectMsg(blueTagWrite)
+      val redTagWrite = TagWrite("red", Vector.empty)
+      tagWriters ! redTagWrite
+      redProbe.expectMsg(redTagWrite)
+
+      tagWriters ! PidRecovering("p1", Map("red" -> TagProgress("p1", 2, 1)))
+      redProbe.expectMsg(ResetPersistenceId("p1", "red", TagProgress("p1", 2, 1)))
+      blueProbe.expectMsg(ResetPersistenceId("p1", "blue", TagProgress("p1", 0, 0)))
+      expectNoMessage(smallWait)
+      redProbe.reply(ResetPersistenceIdComplete)
+      expectNoMessage(smallWait)
+      blueProbe.reply(ResetPersistenceIdComplete)
+      expectMsg(PidRecoveringAck)
+    }
```

Comment:
This one I don't quite understand, what is a blank progress, that the `TagProgress` stays on the same sequence and `TagPidSequenceNr`?

Reply:
It sets it back to as if there was no progress.

This is because if recovery doesn't find any progress in the tag_write_progress table it recovers as if no tag writes ever made it, so the tag writers need to be reset to that state too. This happens if you start a persistent actor, write some messages and restart it before any of the tag writes were written to Cassandra.